src/main/java/org/galatea/starter/service/IexService.java

```diff
@@ -59,10 +72,115 @@
     if (symbols.length() == 0) {
       return Collections.emptyList();
     } else if (date == null) {
-      return newClient.getHistoricalPricesWithRange(symbols, range);
+      checkAndRetrieveForRange(service, symbols, range);
+      return retrieveFromDBForRange(service, symbols, range);
     } else {
-      return newClient.getHistoricalPricesWithDate(symbols, date);
+      checkAndRetrieveForDate(service, symbols, date);
+      return retrieveFromDBForDate(service, symbols, date);
+    }
+  }
+
+  public void checkAndRetrieveForRange(HistoricalPriceDBService service, String symbol,
+      String rangeOfDays) {
+    String r = rangeOfDays.split("(?<=\\D)(?=\\d)|(?<=\\d)(?=\\D)")[0];
+    int range = 30;
+    if (r.length() != 0) {
+      range = Integer.parseInt(r);
+    }
+    LocalDate today = LocalDate.now();
+    for (int i = range; i >= 1; i--) {
+      LocalDate date = today.minusDays(i);
+      DateTimeFormatter formatters = DateTimeFormatter.ofPattern("yyyyMMdd");
+      CompositePrimaryKey obj = new CompositePrimaryKey(symbol, date,
+          LocalTime.now().withHour(0).withMinute(0).withSecond(0).withNano(0));
+      if (!service.exists(obj)) {
+        List<IexHistoricalPrice> prices =
+            newClient.getHistoricalPriceWithDateByDay(symbol, date.format(formatters));
+        for (IexHistoricalPrice price : prices) {
+          LocalTime t = LocalTime.now().withHour(0).withMinute(0).withSecond(0).withNano(0);
```

Comment:
why are we using localTime as part of the key, if we always save it with all values of 0?

Reply:
While making a range query(price of AAPL over 5 days), we get one price for each day. Whereas when we make a query for a particular date, we get prices for each minute from 9:30 AM to 4:00 PM. Since I am using the same table to store prices from both of these queries, I need a way of distinguishing whether this price is from the range query or the date query. If I were to repeat the range query, I would use the entry corresponding to (symbol,date,00:00). If I were to repeat the date query, I would use the entries from (symbol,date,09:30) to (symbol,date,16:00).